File: cs/src/IceSSL/SslTransceiver.cs
// **********************************************************************
//
// Copyright (c) 2003-2006 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

namespace IceSSL
{
    using System.ComponentModel;
    using System.Diagnostics;
    using System.Net.Security;
    using System.Net.Sockets;
    using System.IO;

    sealed class SslTransceiver : IceInternal.Transceiver
    {
	public Socket fd()
	{
	    Debug.Assert(fd_ != null);
	    return fd_;
	}

	public void close()
	{
	    if(instance_.networkTraceLevel() >= 1)
	    {
		string s = "closing ssl connection\n" + ToString();
		logger_.trace(instance_.networkTraceCategory(), s);
	    }

	    lock(this)
	    {
		Debug.Assert(fd_ != null);
		Debug.Assert(stream_ != null);
		try
		{
		    //
		    // Closing the stream also closes the socket.
		    //
		    stream_.Close();
		}
		catch(IOException ex)
		{
		    throw new Ice.SocketException(ex);
		}
		finally
		{
		    fd_ = null;
		    stream_ = null;
		}
	    }
	}

	public void shutdownWrite()
	{
	    if(instance_.networkTraceLevel() >= 2)
	    {
		string s = "shutting down ssl connection for writing\n" + ToString();
		logger_.trace(instance_.networkTraceCategory(), s);
	    }

	    Debug.Assert(fd_ != null);
	    try
	    {
		fd_.Shutdown(SocketShutdown.Send);
	    }
	    catch(SocketException ex)
	    {
	        if(IceInternal.Network.notConnected(ex))
		{
		    return;
		}
	        throw new Ice.SocketException(ex);
	    }
	}

	public void shutdownReadWrite()
	{
	    if(instance_.networkTraceLevel() >= 2)
	    {
		string s = "shutting down ssl connection for reading and writing\n" + ToString();
		logger_.trace(instance_.networkTraceCategory(), s);
	    }

	    Debug.Assert(fd_ != null);
	    try
	    {
		fd_.Shutdown(SocketShutdown.Both);
	    }
	    catch(SocketException ex)
	    {
	        if(IceInternal.Network.notConnected(ex))
		{
		    return;
		}
	        throw new Ice.SocketException(ex);
	    }
	}

	public void write(IceInternal.BasicStream stream, int timeout)
	{
	    Debug.Assert(fd_ != null);

	    /*
	     * TODO: Timeouts
	    if(timeout == -1)
	    {
		timeout = System.Threading.Timeout.Infinite;
	    }
	    if(stream_.WriteTimeout != timeout)
	    {
		stream_.WriteTimeout = timeout;
	    }
	    */

	    IceInternal.ByteBuffer buf = stream.prepareWrite();
	    int remaining = buf.remaining();
	    int position = buf.position();
	    try
	    {
		stream_.Write(buf.rawBytes(), position, remaining);

		if(instance_.networkTraceLevel() >= 3)
		{
		    string s = "sent " + remaining + " of " + remaining + " bytes via ssl\n" + ToString();
		    logger_.trace(instance_.networkTraceCategory(), s);
		}
		if(stats_ != null)
		{
		    stats_.bytesSent(type(), remaining);
		}
	    }
	    catch(IOException ex)
	    {
		// TODO: Is there a better way than checking the message string?
		if(IceInternal.Network.connectionLost(ex))
		{
		    throw new Ice.ConnectionLostException(ex);
		}
		if(IceInternal.Network.timeout(ex))
		{
		    throw new Ice.TimeoutException();
		}
		throw new Ice.SocketException(ex);
	    }
	    catch(SocketException ex) // TODO: Necessary?
	    {
		if(IceInternal.Network.connectionLost(ex))
		{
		    throw new Ice.ConnectionLostException(ex);
		}
		if(IceInternal.Network.wouldBlock(ex))
		{
		    throw new Ice.TimeoutException();
		}
		throw new Ice.SocketException(ex);
	    }
	    catch(Ice.LocalException)
	    {
		throw;
	    }
	    catch(System.Exception ex)
	    {
		throw new Ice.SyscallException(ex);
	    }
	}

	public void read(IceInternal.BasicStream stream, int timeout)
	{
	    Debug.Assert(fd_ != null);

	    /*
	     * TODO: Timeouts
	    if(timeout == -1)
	    {
		timeout = System.Threading.Timeout.Infinite;
	    }
	    if(stream_.ReadTimeout != timeout)
	    {
		stream_.ReadTimeout = timeout;
	    }
	    */

	    //
	    // TODO: Is this necessary to work around a .NET 2.0 bug?
	    //
	    // http://forums.microsoft.com/MSDN/ShowPost.aspx?PostID=171136&SiteID=1
	    //
	    fd_.Blocking = true;

	    IceInternal.ByteBuffer buf = stream.prepareRead();    
	    int remaining = buf.remaining();
	    int position = buf.position();

	    try
	    {
		int ret = stream_.Read(buf.rawBytes(), position, remaining);
		if(ret == 0)
		{
		    //
		    // Try to read again; if zero is returned, the connection is lost.
		    //
		    ret = stream_.Read(buf.rawBytes(), position, remaining);
		    if(ret == 0)
		    {
			throw new Ice.ConnectionLostException();
		    }
		}
		if(instance_.networkTraceLevel() >= 3)
		{
		    string s = "received " + ret + " of " + remaining + " bytes via ssl\n" + ToString();
		    logger_.trace(instance_.networkTraceCategory(), s);
		}    
		if(stats_ != null)
		{
		    stats_.bytesReceived(type(), ret);
		}
		buf.position(position += ret);
	    }
	    catch(IOException ex)
	    {
		// TODO: Is there a better way than checking the message string?
		if(IceInternal.Network.connectionLost(ex))
		{
		    throw new Ice.ConnectionLostException(ex);
		}
		if(IceInternal.Network.timeout(ex))
		{
		    throw new Ice.TimeoutException();
		}
		throw new Ice.SocketException(ex);
	    }
	    catch(SocketException ex) // TODO: Necessary?
	    {
		if(IceInternal.Network.connectionLost(ex))
		{
		    throw new Ice.ConnectionLostException(ex);
		}
		if(IceInternal.Network.wouldBlock(ex))
		{
		    throw new Ice.TimeoutException();
		}
		throw new Ice.SocketException(ex);
	    }
	    catch(Ice.LocalException) // TODO: Necessary?
	    {
		throw;
	    }
	    catch(System.Exception ex)
	    {
		throw new Ice.SyscallException(ex);
	    }
	}

	public string type()
	{
	    return "ssl";
	}

	public override string ToString()
	{
	    return desc_;
	}

	//
	// Only for use by SslConnector, SslAcceptor.
	//
	internal SslTransceiver(Instance instance, Socket fd, SslStream stream)
	{
	    instance_ = instance;
	    fd_ = fd;
	    stream_ = stream;
	    logger_ = instance.communicator().getLogger();
	    stats_ = instance.communicator().getStats();
	    desc_ = IceInternal.Network.fdToString(fd_);
	}

#if DEBUG
	~SslTransceiver()
	{
	    lock(this)
	    {
		IceUtil.Assert.FinalizerAssert(fd_ == null);
	    }
	}
#endif

	private Instance instance_;
	private Socket fd_;
	private SslStream stream_;
	private Ice.Logger logger_;
	private Ice.Stats stats_;
	private string desc_;
    }

}

File: java/src/IceSSL/SslTransceiver.java
// **********************************************************************
//
// Copyright (c) 2003-2006 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

package IceSSL;

final class SslTransceiver implements IceInternal.Transceiver
{
    public java.nio.channels.SelectableChannel
    fd()
    {
	return null;
    }

    public void
    close()
    {
	if(_instance.networkTraceLevel() >= 1)
	{
	    String s = "closing ssl connection\n" + toString();
	    _logger.trace(_instance.networkTraceCategory(), s);
	}

	synchronized(this)
	{
	    assert(_fd != null);
	    try
	    {
		_fd.close();
	    }
	    catch(java.io.IOException ex)
	    {
		Ice.SocketException se = new Ice.SocketException();
		se.initCause(ex);
		throw se;
	    }
	    finally
	    {
		_fd = null;
	    }
	}
    }

    public void
    shutdownWrite()
    {
	/*
	 * shutdownOutput is not supported by an SSL socket.
	 *
	if(_instance.networkTraceLevel() >= 2)
	{
	    String s = "shutting down ssl connection for writing\n" + toString();
	    _logger.trace(_instance.networkTraceCategory(), s);
	}

	assert(_fd != null);
	try
	{
	    _fd.shutdownOutput(); // Shutdown socket for writing
	}
	catch(UnsupportedOperationException ex)
	{
	    // Ignore - shutdownOutput not supported.
	}
	catch(java.io.IOException ex)
	{
	    Ice.SocketException se = new Ice.SocketException();
	    se.initCause(ex);
	    throw se;
	}
	*/
    }

    public void
    shutdownReadWrite()
    {
	if(_instance.networkTraceLevel() >= 2)
	{
	    String s = "shutting down ssl connection for reading and writing\n" + toString();
	    _logger.trace(_instance.networkTraceCategory(), s);
	}

	assert(_fd != null);

	_shutdown = true;

	/*
	 * shutdownInput is not supported by an SSL socket.
	 *
	try
	{
	    _fd.shutdownInput(); // Shutdown socket for reading
	    //_fd.shutdownOutput(); // Shutdown socket for writing
	}
	catch(UnsupportedOperationException ex)
	{
	    // Ignore - shutdownInput not supported.
	}
	catch(java.net.SocketException ex)
	{
	    // Ignore.
	}
	catch(java.io.IOException ex)
	{
	    Ice.SocketException se = new Ice.SocketException();
	    se.initCause(ex);
	    throw se;
	}
	*/
    }

    public void
    write(IceInternal.BasicStream stream, int timeout)
    {
	java.nio.ByteBuffer buf = stream.prepareWrite();

	byte[] data = null;
	int off = 0;
	try
	{
	    data = buf.array();
	    off = buf.arrayOffset();
	}
	catch(UnsupportedOperationException ex)
	{
	    assert(false);
	}

	try
	{
	    if(timeout == -1)
	    {
		timeout = 0; // Infinite
	    }
	    else if(timeout == 0)
	    {
		timeout = 1;
	    }
	    _fd.setSoTimeout(timeout);
	}
	catch(java.net.SocketException ex)
	{
	    Ice.SocketException se = new Ice.SocketException();
	    se.initCause(ex);
	    throw se;
	}

	while(buf.hasRemaining())
	{
	    int pos = buf.position();
	    try
	    {
		assert(_fd != null);
		int rem = buf.remaining();
		_out.write(data, off + pos, rem);
		buf.position(pos + rem);

		if(_instance.networkTraceLevel() >= 3)
		{
		    String s = "sent " + rem + " of " + buf.limit() + " bytes via ssl\n" + toString();
		    _logger.trace(_instance.networkTraceCategory(), s);
		}

		if(_stats != null)
		{
		    _stats.bytesSent(type(), rem);
		}

		break;
	    }
	    catch(java.io.InterruptedIOException ex)
	    {
		buf.position(pos + ex.bytesTransferred);
	    }
	    catch(java.io.IOException ex)
	    {
		Ice.SocketException se = new Ice.SocketException();
		se.initCause(ex);
		throw se;
	    }
	}
    }

    public void
    read(IceInternal.BasicStream stream, int timeout)
    {
	java.nio.ByteBuffer buf = stream.prepareRead();

	int remaining = 0;
	if(_instance.networkTraceLevel() >= 3)
	{
	    remaining = buf.remaining();
	}

	byte[] data = null;
	int off = 0;
	try
	{
	    data = buf.array();
	    off = buf.arrayOffset();
	}
	catch(UnsupportedOperationException ex)
	{
	    assert(false);
	}

	int interval = 500;
	if(timeout >= 0 && timeout < interval)
	{
	    interval = timeout;
	}

	while(buf.hasRemaining() && !_shutdown)
	{
	    int pos = buf.position();
	    try
	    {
		_fd.setSoTimeout(interval);
		assert(_fd != null);
		int ret = _in.read(data, off + pos, buf.remaining());
		
		if(ret == -1)
		{
		    throw new Ice.ConnectionLostException();
		}

		if(ret > 0)
		{
		    if(_instance.networkTraceLevel() >= 3)
		    {
			String s = "received " + ret + " of " + remaining + " bytes via ssl\n" + toString();
			_logger.trace(_instance.networkTraceCategory(), s);
		    }

		    if(_stats != null)
		    {
			_stats.bytesReceived(type(), ret);
		    }

		    buf.position(pos + ret);
		}
	    }
	    catch(java.net.SocketTimeoutException ex)
	    {
		if(ex.bytesTransferred > 0)
		{
		    buf.position(pos + ex.bytesTransferred);
		}
		if(timeout >= 0)
		{
		    if(interval >= timeout)
		    {
			throw new Ice.TimeoutException();
		    }
		    timeout -= interval;
		}
	    }
	    catch(java.io.InterruptedIOException ex)
	    {
		buf.position(pos + ex.bytesTransferred);
	    }
	    catch(java.io.IOException ex)
	    {
		if(IceInternal.Network.connectionLost(ex))
		{
		    Ice.ConnectionLostException se = new Ice.ConnectionLostException();
		    se.initCause(ex);
		    throw se;
		}
		
		Ice.SocketException se = new Ice.SocketException();
		se.initCause(ex);
		throw se;
	    }
	}

	if(_shutdown)
	{
	    throw new Ice.ConnectionLostException();
	}
    }

    public String
    type()
    {
	return "ssl";
    }

    public String
    toString()
    {
	return _desc;
    }

    //
    // Only for use by SslConnector, SslAcceptor
    //
    SslTransceiver(Instance instance, javax.net.ssl.SSLSocket fd)
    {
	_instance = instance;
	_fd = fd;
	_logger = instance.communicator().getLogger();
	try
	{
	    _stats = instance.communicator().getStats();
	}
	catch(Ice.CommunicatorDestroyedException ex)
	{
	    // Ignore.
	}
	_desc = IceInternal.Network.fdToString(_fd);
	try
	{
	    _in = _fd.getInputStream();
	    _out = _fd.getOutputStream();
	}
	catch(java.io.IOException ex)
	{
	    try
	    {
		_fd.close();
	    }
	    catch(java.io.IOException e)
	    {
	    }
	    _fd = null;
	    Ice.SocketException se = new Ice.SocketException();
	    se.initCause(ex);
	    throw se;
	}
	_shutdown = false;
    }

    protected synchronized void
    finalize()
	throws Throwable
    {
	assert(_fd == null);

	super.finalize();
    }

    private Instance _instance;
    private javax.net.ssl.SSLSocket _fd;
    private Ice.Logger _logger;
    private Ice.Stats _stats;
    private String _desc;
    private java.io.InputStream _in;
    private java.io.OutputStream _out;
    private volatile boolean _shutdown;
}
